File: YACS Python Solution/2021.1 YACS 丙组乙组甲组/YACS 2021.1 乙组 #2.py
n,m,s = map(int, input().split()) # n名运动员，m体力，s总圈数, s<=m

value = [m]*n
result = []

def change(cur, s_remain, time_taken):
    if s_remain==0:
        result.append(time_taken)
        return
    max_v = max(cur)
    for speed in range(1,s+1):
        if max_v<speed**2 or s_remain - speed<0: break
        new = [x-speed for x in cur]
        for j in range(n):
            if new[j] == max_v-speed:
                new[j] = new[j]+speed-speed**2
                break
        change(new, s_remain - speed, time_taken+1)

change(value, s, 0)
print(min(result))

File: YACS Python Solution/2021.01 YACS 丙组乙组甲组/YACS 2021.1 乙组 #2.cpp
#include <bits/stdc++.h>
using namespace std;
int n,m,s, ans=99999;
vector<int> results;

int maxlist(int tmp[51]){
    int tmpans=0;
    for(int i=0;i< n; i++) tmpans = max(tmpans,tmp[i]);
    return tmpans;
}

void dfs(int power[51],int circles, int tt)
{
    // tt: time_taken
    if (circles==0){ans = min(ans, tt); return;}

    int max_v = maxlist(power);

    for(int speed=1;speed<=n;speed++){
        if (max_v<(speed*speed) || (circles-speed) < 0) break;
        
        int newlist[51];
        for(int i=0;i< n; i++) newlist[i] = power[i]-speed;

        for(int i=0;i< n; i++){
            if (newlist[i] == max_v-speed){ 
                newlist[i] = newlist[i]+speed-speed*speed; 
                break;
            }
        }
        dfs(newlist, circles-speed, tt+1);
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin>>n>>m>>s;

    int value[51];
    for (int i=0;i<n;i++) value[i]=m;
    dfs(value,s,0);

    cout<<ans<<endl;
    return 0;
}